Extension/src/LanguageServer/extension.ts

```diff
@@ -60,6 +63,72 @@ export function activate(activationEventOccurred: boolean): void {
     }
 }
 
+export function registerCustomConfigurationProvider(provider: CustomConfigurationProvider): void {
+    customConfigurationProviders.push(provider);
+}
+
+export async function provideCustomConfiguration(document: vscode.TextDocument, client: Client): Promise<void> {
+    let tokenSource: CancellationTokenSource = new CancellationTokenSource();
+    return runBlockingThenableWithTimeout(async () => {
+        // Loop through registered providers until one is able to service the current document
+        for (let i: number = 0; i < customConfigurationProviders.length; i++) {
+            if (await customConfigurationProviders[i].canProvideConfiguration(document.uri)) {
+                return customConfigurationProviders[i].provideConfigurations([document.uri]);
+            }
+        }
+        return Promise.reject("No providers found for " + document.uri);
+    }, 1000, client, tokenSource).then((configs: SourceFileConfigurationItem[]) => {
+        if (configs !== null && configs.length > 0) {
+            client.sendCustomConfigurations(configs);
+        }
+    });
+}
+
+export function onDidCustomConfigurationChange(customConfigurationProvider: CustomConfigurationProvider): void {
+    // TODO: How to handle multiple workspaces? Only works with the current workspace.
+    // Should this be called when the workspace changes?
+    if (vscode.workspace.textDocuments === undefined || vscode.workspace.textDocuments.length === 0) {
+        return;
+    }
+
+    let documents: vscode.Uri[] = [];
+    for (let i: number = 0; i < vscode.workspace.textDocuments.length; ++i) {
+        let document: vscode.TextDocument = vscode.workspace.textDocuments[i];
+        if (document.languageId === "cpp" || document.languageId === "c") {
+            documents.push(document.uri);
+        }
+    }
+
+    let tokenSource: CancellationTokenSource = new CancellationTokenSource();
+
+    // TODO: Can we rely on clients.ActiveClient to be the client for the current workspace?
+    runBlockingThenableWithTimeout(() => customConfigurationProvider.provideConfigurations(documents, tokenSource.token), 1000, clients.ActiveClient, tokenSource)
+    .then((configs: SourceFileConfigurationItem[]) => {
+        clients.ActiveClient.sendCustomConfigurations(configs);
+    });
+}
+
+function runBlockingThenableWithTimeout(thenable: () => Thenable<any>, ms: number, client: Client, tokenSource?: CancellationTokenSource): Thenable<any> {
+    let timer: NodeJS.Timer;
+    // Create a promise that rejects in <ms> milliseconds
+    let timeout: Promise<any> = new Promise((resolve, reject) => {
+        timer = setTimeout(() => {
+            clearTimeout(timer);
+            if (tokenSource) {
+                tokenSource.cancel();
```

Comment:
I see you are marking the operation as cancelled, but I don't see any code that actually does anything about it.  Are there future changes coming for this?

Reply:
The token is passed to our extension, and then passed to our "configuration server", which will react to it. Is there something to be done here as well?

Reply:
Not sure how I missed that.  I see it now.